crates/catalog/sql/src/catalog.rs

```diff
@@ -276,4 +579,385 @@ mod tests {
+
+        catalog.drop_namespace(&namespace_ident_a_b).await.unwrap();
+
+        assert!(!catalog
+            .namespace_exists(&namespace_ident_a_b)
+            .await
+            .unwrap());
+
+        assert!(catalog.namespace_exists(&namespace_ident_a).await.unwrap());
+    }
+
+    #[tokio::test]
+    async fn test_drop_deeply_nested_namespace() {
+        let warehouse_loc = temp_path();
+        let catalog = new_sql_catalog(warehouse_loc).await;
+        let namespace_ident_a = NamespaceIdent::new("a".into());
+        let namespace_ident_a_b = NamespaceIdent::from_strs(vec!["a", "b"]).unwrap();
+        let namespace_ident_a_b_c = NamespaceIdent::from_strs(vec!["a", "b", "c"]).unwrap();
+        create_namespaces(&catalog, &vec![
+            &namespace_ident_a,
+            &namespace_ident_a_b,
+            &namespace_ident_a_b_c,
+        ])
+        .await;
+
+        catalog
+            .drop_namespace(&namespace_ident_a_b_c)
+            .await
+            .unwrap();
+
+        assert!(!catalog
+            .namespace_exists(&namespace_ident_a_b_c)
+            .await
+            .unwrap());
+
+        assert!(catalog
+            .namespace_exists(&namespace_ident_a_b)
+            .await
+            .unwrap());
+
+        assert!(catalog.namespace_exists(&namespace_ident_a).await.unwrap());
+    }
+
+    #[tokio::test]
+    async fn test_drop_namespace_throws_error_if_namespace_doesnt_exist() {
+        let warehouse_loc = temp_path();
+        let catalog = new_sql_catalog(warehouse_loc).await;
+
+        let non_existent_namespace_ident = NamespaceIdent::new("abc".into());
+        assert_eq!(
+            catalog
+                .drop_namespace(&non_existent_namespace_ident)
+                .await
+                .unwrap_err()
+                .to_string(),
+            format!(
+                "Unexpected => No such namespace: {:?}",
+                non_existent_namespace_ident
+            )
+        )
+    }
+
+    #[tokio::test]
+    async fn test_drop_namespace_throws_error_if_nested_namespace_doesnt_exist() {
+        let warehouse_loc = temp_path();
+        let catalog = new_sql_catalog(warehouse_loc).await;
+        create_namespace(&catalog, &NamespaceIdent::new("a".into())).await;
+
+        let non_existent_namespace_ident =
+            NamespaceIdent::from_vec(vec!["a".into(), "b".into()]).unwrap();
+        assert_eq!(
+            catalog
+                .drop_namespace(&non_existent_namespace_ident)
+                .await
+                .unwrap_err()
+                .to_string(),
+            format!(
+                "Unexpected => No such namespace: {:?}",
+                non_existent_namespace_ident
+            )
+        )
+    }
+
+    #[tokio::test]
+    #[ignore = "Java/Python do not drop nested namespaces?"]
+    async fn test_dropping_a_namespace_also_drops_namespaces_nested_under_that_one() {
```

Comment:
I left this as an ignored test because I think this is up for debate. It's technically correct that we should not allow dropping a namespace with child namespaces, but the other implementations do not have this.

Maybe as an alternative I could implement a check to fail when there are child namespaces in a call to `drop_namespace`? What do you think?

Reply:
I think your second suggestion, only allowing dropping a namespace when it is empty, makes sense. Or how about adding a boolean flag, `drop_children`? If true, drop any children present in the namespace as well as the namespace. If false, only drop the namespace if it is empty, and return an `Err` otherwise.

All of these approaches seem more sensible than the Python / Java approach of dropping a namespace but leaving any tables that were within it with a missing parent. That feels buggy.

Reply:
I will add this when I re-add the drop_namespaces function in the next PR for tables. I'll check if the namespace is empty for both tables, and child namespaces.